### PR TITLE
Remove guava dependency

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
Looks like now everybody will do this)

### CODE REVIEW

**Comment by AlexanderYastrebov** [comment]
Fixes #59

**Comment by whiskeysierra** [comment]
Ah! Ok...

File: logbook-core/src/main/java/org/zalando/logbook/DefaultLogbook.java [hunk 1]
```diff
@@ -70,7 +68,7 @@
         }
     }
 
-    @VisibleForTesting
+    //@VisibleForTesting
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultLogbook.java]: I'd just drop it

File: logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java [hunk 2]
```diff
@@ -42,15 +42,15 @@ public Logbook create(
             @Nullable final HttpLogWriter writer) {
 
 
-        final HeaderObfuscator header = firstNonNull(headerObfuscator, authorization());
-        final BodyObfuscator body = firstNonNull(bodyObfuscator, BodyObfuscator.none());
+        final HeaderObfuscator header = Optional.ofNullable(headerObfuscator).orElse(authorization());
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java]: Could use `orElseGet(Obfuscators::authorization)`

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java [hunk 3]
```diff
@@ -59,20 +55,18 @@ public String format(final Precorrelation<HttpRequest> precorrelation) throws IO
         final String correlationId = precorrelation.getId();
         final HttpRequest request = precorrelation.getRequest();
 
-        final ImmutableMap.Builder<String, Object> builder = ImmutableMap.<String, Object>builder();
+        final Map<String, Object> content = new HashMap<>();
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: I'd use a LinkedHashMap

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java [hunk 4]
```diff
@@ -138,12 +130,12 @@ private boolean isJson(final String contentType) {
...
-        final MediaType mediaType = MediaType.parse(contentType);
+        final String t = contentType.toLowerCase();
 
-        final boolean isJson = mediaType.is(APPLICATION_JSON);
+        final boolean isJson = t.startsWith("application/json");
 
-        final boolean isApplication = mediaType.is(MediaType.ANY_APPLICATION_TYPE);
-        final boolean isCustomJson = mediaType.subtype().endsWith("+json");
+        final boolean isApplication = t.startsWith("application/");
+        final boolean isCustomJson = t.endsWith("+json");
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: It's missing support for parameters, e.g. `application/custom+json;version=2`

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java [hunk 5]
```diff
@@ -174,7 +166,7 @@ private String compactJson(final String json) throws IOException {
 
     // this wouldn't catch spaces in json, but that's ok for our use case here
     private boolean isAlreadyCompacted(final String json) {
-        return PRETTY_PRINT.matchesNoneOf(json);
+        return (json.indexOf('\n') == -1);
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Parenthesis are not needed.

File: logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java [hunk 6]
```diff
@@ -130,32 +131,36 @@ public void headerShouldNotMatchNameAndValue() {
...
     }
 
-    @Test
-    public void headerShouldNotMatchPredicate() {
-        final Predicate<RawHttpRequest> unit = header((name, value) -> 
-                name.equalsIgnoreCase("X-Secret") && value.equalsIgnoreCase("false"));
-
-        assertThat(unit.test(request), is(false));
-    }
+//    @Test
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java]: What happened here?

File: logbook-core/src/test/java/org/zalando/logbook/ObfuscatorsTest.java [hunk 7]
```diff
@@ -59,4 +63,15 @@ public void authorizationShouldNotObfuscateNonAuthorization() {
         assertThat(unit.obfuscate("Accept", "text/plain"), is(equalTo("text/plain")));
     }
 
-}
\ No newline at end of file
+    @Test
+    public void shouldObfuscateMap() {
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/ObfuscatorsTest.java]: The test is not verifying that it actually obfuscates.

File: logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+ /*
+ * ⁣​
+ * Jackson-datatype-Problem
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java]: Copy and paste?

File: logbook-servlet/src/test/java/org/zalando/logbook/servlet/ForwardingHttpLogFormatter.java [hunk 9]
```diff
@@ -53,4 +51,8 @@ public String format(final Correlation<HttpRequest, HttpResponse> correlation) t
         return delegate().format(correlation);
     }
 
+    @Override
+    public String toString() {
+        return delegate().toString();
```
Comment (whiskeysierra) [inline logbook-servlet/src/test/java/org/zalando/logbook/servlet/ForwardingHttpLogFormatter.java]: Good catch! I totally forgot that.

File: logbook-test/src/test/java/org/zalando/logbook/MockHeadersTest.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+ /*
```
Comment (whiskeysierra) [inline logbook-test/src/test/java/org/zalando/logbook/MockHeadersTest.java]: Double headers?

**Comment by whiskeysierra** [comment]
Almost scary how similar our code is :ghost:

**Comment by whiskeysierra** [comment]
:+1:

File: logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java [hunk 11]
```diff
@@ -35,22 +35,42 @@
...
 
     Charset getCharset();
 
-    class Headers {
+    class HeadersBuilder {
+
+        private final Map<String, List<String>> headers;
 
-        Headers() {
+        public HeadersBuilder() {
```
Comment (whiskeysierra) [inline logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java]: Could be its own class now.

File: logbook-core/src/main/java/org/zalando/logbook/Conditions.java [hunk 12]
```diff
@@ -70,18 +71,18 @@
...
         return request ->
                 request.getHeaders().get(key).stream().anyMatch(predicate);
     }
 
-    public static Predicate<RawHttpRequest> header(final BiPredicate<String, String> predicate) {
-        return request ->
-                request.getHeaders().entries().stream()
-                        .anyMatch(e -> predicate.test(e.getKey(), e.getValue()));
-    }
+//    public static Predicate<RawHttpRequest> header(final BiPredicate<String, String> predicate) {
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/Conditions.java]: Why the comments?

File: logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java (hunk 2 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java]: Same goes for the other places where the alternative is a method call.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 4 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Not 100% correct, but I'd go with it.

```java
final boolean isCustomJson = t.endsWith("+json") || t.contains("+json;");
```

File: logbook-core/src/main/java/org/zalando/logbook/Obfuscators.java [hunk 13]
```diff
@@ -56,4 +58,12 @@ public static HeaderObfuscator authorization() {
         return obfuscate("Authorization"::equalsIgnoreCase, "XXX");
     }
 
+    static Map<String, List<String>> obfuscateHeaders(final Map<String, List<String>> map, final BiFunction<String, String, String> f) {
+        final BaseHttpMessage.HeadersBuilder builder = new BaseHttpMessage.HeadersBuilder();
+        for (Map.Entry<String, List<String>> e : map.entrySet()) {
+            final String k = e.getKey();
+            builder.put(k, e.getValue().stream().map(x -> f.apply(k, x)).collect(Collectors.toList()));
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/Obfuscators.java]: Static import for `toList`?

File: logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java [hunk 14]
```diff
@@ -130,32 +131,36 @@ public void headerShouldNotMatchNameAndValue() {
...
         assertThat(unit.test(request), is(false));
     }
 
-    @Test
-    public void headerShouldMatchPredicate() {
-        final Predicate<RawHttpRequest> unit = header((name, value) -> 
-                name.equalsIgnoreCase("X-Secret") && value.equalsIgnoreCase("true"));
-
-        assertThat(unit.test(request), is(true));
+    private static Set<String> setOf(String... values) {
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java]: :+1:

File: logbook-core/src/test/java/org/zalando/logbook/EnforceCoverageTest.java [hunk 15]
```diff
@@ -20,7 +20,7 @@
  * #L%
  */
 
-/*
+ /*
  * ⁣​
  * Jackson-datatype-Problem
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/EnforceCoverageTest.java]: Copy and paste?

File: logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
+package org.zalando.logbook.io;
+
+/*
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java]: Double header?

File: logbook-test/src/main/java/org/zalando/logbook/MockHeaders.java [hunk 17]
```diff
@@ -20,18 +20,47 @@
  * #L%
  */
 
-import com.google.common.collect.ListMultimap;
+/*
+ * Copyright 2016 Zalando SE.
```
Comment (whiskeysierra) [inline logbook-test/src/main/java/org/zalando/logbook/MockHeaders.java]: Double header?

File: logbook-test/src/main/java/org/zalando/logbook/MockHeaders.java [hunk 18]
```diff
@@ -2,7 +2,7 @@
 
 /*
  * #%L
- * Logbook: Test
```
Comment (whiskeysierra) [inline logbook-test/src/main/java/org/zalando/logbook/MockHeaders.java]: Should be Test, based on the file location

File: logbook-core/src/main/java/org/zalando/logbook/Conditions.java (hunk 12 above)
Comment (AlexanderYastrebov) [inline logbook-core/src/main/java/org/zalando/logbook/Conditions.java]: Did not find a straightforward way to implement this so just dropped for now)

File: logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java (hunk 11 above)
Comment (AlexanderYastrebov) [inline logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java]: I am thinking of implementing Header and Headers implements Iterable<Header> classes instead of map of lists as this will allow to extend it in future.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 3 above)
Comment (AlexanderYastrebov) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: But what for? I am not sure it will serialize in insertion order.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 5 above)
Comment (AlexanderYastrebov) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Sure. It is just for visual grouping.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 3 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Yes, it will

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 5 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: But there is only one group... :point_up:

File: logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java (hunk 11 above)
Comment (whiskeysierra) [inline logbook-api/src/main/java/org/zalando/logbook/BaseHttpMessage.java]: But we need a map for easy lookups, e.g. in conditions.

File: logbook-core/src/main/java/org/zalando/logbook/Conditions.java (hunk 12 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/Conditions.java]: How about this?

```java
return request ->
        request.getHeaders().entrySet().stream()
                .anyMatch(e -> e.getValue().stream().anyMatch(v -> predicate.test(e.getKey(), v)));
```

File: logbook-core/src/main/java/org/zalando/logbook/Conditions.java [hunk 19]
```diff
@@ -70,18 +71,18 @@
 
     public static Predicate<RawHttpRequest> header(final String key, final String value) {
         return request ->
-                request.getHeaders().containsEntry(key, value);
+                request.getHeaders().getOrDefault(key, Collections.emptyList()).contains(value);
     }
 
     public static Predicate<RawHttpRequest> header(final String key, final Predicate<String> predicate) {
         return request ->
                 request.getHeaders().get(key).stream().anyMatch(predicate);
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/Conditions.java]: This will throw an exception now if a key doesn't exist. Can you verify with a test?

File: logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        long total = 0;
+        while (true) {
+            int r = from.read(buf);
+            if (r == -1) {
+                break;
+            }
+            to.write(buf, 0, r);
+            total += r;
+        }
+        return total;
```
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java]: We don't need the total anywhere

File: logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java (hunk 14 above)
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/ConditionsTest.java]: Although asList woulf also work. Since we are only using it in tests.

File: logbook-core/src/test/java/org/zalando/logbook/ObfuscatorsTest.java [hunk 21]
```diff
@@ -59,4 +64,17 @@ public void authorizationShouldNotObfuscateNonAuthorization() {
...
-}
\ No newline at end of file
+    @Test
+    public void shouldObfuscateHeaders() {
+        final Map<String, List<String>> m = Obfuscators.obfuscateHeaders(
+            MockHeaders.of("hello", "world", "chao", "bambina"),
+            (k, v) -> v.toUpperCase()
+        );
+
+        assertEquals(2, m.size());
```
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/ObfuscatorsTest.java]:
```java
assertThat(m, hasSize(2));
assertThat(m, hasKey("hello"));
assertThat(m, hasKey("chao"));
assertThat(m, hasEntry(equalTo("hello"), contains("WORLD")));
assertThat(m, hasEntry(equalTo("chao"), contains("BAMBINA")));
```

File: logbook-httpclient/src/test/java/org/zalando/logbook/httpclient/EnforceCoverageTest.java [hunk 22]
```diff
@@ -20,26 +20,6 @@
  * #L%
  */
 
-/*
- * ⁣​
- * Jackson-datatype-Problem
```
Comment (whiskeysierra) [inline logbook-httpclient/src/test/java/org/zalando/logbook/httpclient/EnforceCoverageTest.java]: I guess I copied them from somewhere else before...

File: logbook-servlet/src/test/java/org/zalando/logbook/servlet/FormattingTest.java [hunk 23]
```diff
@@ -87,13 +88,13 @@ public void shouldFormatRequest() throws Exception {
...
         final HttpRequest request = interceptRequest();
-        
+
         assertThat(request, hasFeature("remote address", HttpRequest::getRemote, is("127.0.0.1")));
         assertThat(request, hasFeature("method", HttpRequest::getMethod, is("GET")));
         assertThat(request, hasFeature("url", HttpRequest::getRequestUri,
                 hasToString("http://localhost/api/sync?limit=1")));
         assertThat(request, hasFeature("parameters", HttpRequest::getQuery, is("limit=1")));
-        assertThat(request, hasFeature("headers", HttpRequest::getHeaders, is(of("Accept", "text/plain"))));
+        assertThat(request, hasFeature("headers", HttpRequest::getHeaders, is(Collections.singletonMap("Accept", Collections.singletonList("text/plain")))));
```
Comment (whiskeysierra) [inline logbook-servlet/src/test/java/org/zalando/logbook/servlet/FormattingTest.java]: Static imports?

File: logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java (hunk 20 above)
Comment (AlexanderYastrebov) [inline logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java]: I would leave it as is so somebody can use. Overhead is negligible

File: logbook-core/src/main/java/org/zalando/logbook/DefaultHttpLogFormatter.java [hunk 24]
```diff
@@ -59,19 +58,19 @@ private String formatStatusLine(final HttpResponse response) {
     private <H extends HttpMessage> String format(final H message, final String type, final String correlationId,
             final Function<H, String> lineCreator)
             throws IOException {
-        final List<String> lines = Lists.newArrayListWithExpectedSize(4);
+        final List<String> lines = new ArrayList<>();
```
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultHttpLogFormatter.java]: Could still give the expected capacity (which should be 5 instead of 4 to avoid resizing). Indeed for performance reasons this should use a StringBuilder which could be passed to the helper methods. Could be a separate PR though.

File: logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java [hunk 25]
```diff
@@ -42,15 +40,15 @@ public Logbook create(
...
+        final BodyObfuscator body = Optional.ofNullable(bodyObfuscator).orElseGet(BodyObfuscator::none);
 
         return new DefaultLogbook(
-                firstNonNull(condition, $ -> true),
+                Optional.ofNullable(condition).orElse($ -> true),
                 combine(queryObfuscator, header, body, requestObfuscator),
                 combine(header, body, responseObfuscator),
-                firstNonNull(formatter, new DefaultHttpLogFormatter()),
-                firstNonNull(writer, new DefaultHttpLogWriter())
+                Optional.ofNullable(formatter).orElseGet(DefaultHttpLogFormatter::new),
```
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultLogbookFactory.java]: Nice, looks like previously we always unnecessarily created the default instances.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java [hunk 26]
```diff
@@ -82,17 +76,15 @@ public String format(final Correlation<HttpRequest, HttpResponse> correlation) t
...
-        final ImmutableMap.Builder<String, Object> builder = ImmutableMap.<String, Object>builder();
-
-        builder.put("origin", translate(response.getOrigin()));
-        builder.put("type", "response");
-        builder.put("correlation", correlationId);
-        builder.put("protocol", response.getProtocolVersion());
-        builder.put("status", response.getStatus());
-        addUnless(builder, "headers", response.getHeaders().asMap(), Map::isEmpty);
-        addBody(response, builder);
+        final Map<String, Object> content = new HashMap<>();
```
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Should also be `LinkedHashMap`

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java [hunk 27]
```diff
@@ -133,19 +125,20 @@ private Object tryParseBodyAsJson(final String body) {
...
 
-        final MediaType mediaType = MediaType.parse(contentType);
+        final int pi = type.indexOf(';');
+        final int limit = (pi != -1 ? pi : type.length());
 
-        final boolean isJson = mediaType.is(APPLICATION_JSON);
-
-        final boolean isApplication = mediaType.is(MediaType.ANY_APPLICATION_TYPE);
-        final boolean isCustomJson = mediaType.subtype().endsWith("+json");
+        return type.regionMatches(true, 0, "application/json", 0, limit)
```
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Might be more readable as a regular expression like

    final Pattern MEDIA_TYPE_JSON_PATTERN = Pattern.compile("application/(?:json|.+\\+json)(?:$|\\s|;)");

    return MEDIA_TYPE_JSON_PATTERN.matcher(type).lookingAt();

File: logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.OutputStream;
+import java.util.Objects;
+
+public final class ByteStreams {
+
+    ByteStreams() {
+        // package private so we can trick code coverage
+    }
+
+    private static final int BUF_SIZE = 0x1000; // 4K
```
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/io/ByteStreams.java]: Just write 4096 and remove the comment

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 27 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: I haven't tried it, but I think the custom sub type should be non-greedy

File: logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java [hunk 29]
```diff
@@ -22,15 +22,23 @@
...
 
     @Test
-    public void shouldUseMockHeadersConstructor() {
-        new MockHeaders();
+    public void shouldUseByteStreamsConstructor() {
+        new ByteStreams();
     }
 
+    @Test
+    public void shouldCoverByteStreams() throws IOException {
```
Comment (jhorstmann) [inline logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java]: This should be covered by normal tests, it's not to enforce coverage of untestable methods.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 27 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: According to regex101.com it would incorrectly match `application/xml; incompatible=+json`

File: logbook-core/src/main/java/org/zalando/logbook/DefaultHttpLogFormatter.java (hunk 24 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/DefaultHttpLogFormatter.java]: Default is 10, I guess it wouldn't make much of a difference.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 27 above)
Comment (jhorstmann) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: > According to regex101.com it would incorrectly match application/xml; incompatible=+json

Good catch. The definition of tokens and delimiters is here https://tools.ietf.org/html/rfc7230#section-3.2.6 but I think matching everything except space and semicolon should work.

    private static final Pattern MEDIA_TYPE_JSON_PATTERN = Pattern.compile("application/(?:json|[^\t ;]+\\+json)(?:$|\t| |;)");

Comment (AlexanderYastrebov) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: But this one does not look more readable))

File: logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java (hunk 29 above)
Comment (AlexanderYastrebov) [inline logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java]: I copied ByteStreams methods literaly from guava, but this jacoco enforces 100 coverage that is why it is as is.

File: logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java (hunk 27 above)
Comment (whiskeysierra) [inline logbook-core/src/main/java/org/zalando/logbook/JsonHttpLogFormatter.java]: Compared to this?

```java
    private boolean isJson(final String type) {
        if (type.isEmpty()) {
            return false;
        }

        final int pi = type.indexOf(';');
        final int limit = (pi != -1 ? pi : type.length());

        return type.regionMatches(true, 0, "application/json", 0, limit)
            || (type.startsWith("application/") && endsWith(type, "+json", limit));
    }

    private boolean endsWith(final String target, final String what, int limit) {
        return target.regionMatches(true, limit - what.length(), what, 0, what.length());
    }
```

I'd say it's still very competitive.

File: logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java (hunk 29 above)
Comment (whiskeysierra) [inline logbook-core/src/test/java/org/zalando/logbook/io/EnforceCoverageTest.java]: Granted, but it's still a valid test beyond enforcing code coverage. Of course, one could always argue that the only reason we have tests in the first place is to satisfy the coverage requirement.

**Comment by whiskeysierra** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:  nice collective effort

**Comment by AlexanderYastrebov** [comment]
+1

**Comment by AlexanderYastrebov** [comment]
:+1:
